O365Poller/liststate.js

```diff
@@ -0,0 +1,147 @@
+/* -----------------------------------------------------------------------------
+ * @copyright (C) 2017, Alert Logic, Inc
+ * @doc
+ * 
+ * The module for getting storage account settings.
+ * 
+ * @end
+ * -----------------------------------------------------------------------------
+ */
+
+const async = require('async');
+const util = require('util');
+const moment = require('moment');
+const parse = require('parse-key-value');
+
+const m_alUtil = require('../lib/al_util');
+
+var azureStorage = require('azure-storage');
+var base64Encoder = new azureStorage.QueueMessageEncoder.TextBase64QueueMessageEncoder();
+var QueueService = null;
+
+var _getStateQueueName = function() {
+    return 'alertlogic-o365-list';
+};
+
+var _getNewListState = function() {
+    const selectedStreams = process.env.O365_CONTENT_STREAMS;
+    var newState = [];
+    const newCollectedTs = moment.utc().format();
+    
+    for (var i=0; i < selectedStreams.length; ++i) {
+        var newStreamState = {
+            streamName : selectedStreams[i],
+            lastCollectedTs : newCollectedTs
+        };
+        newState.push(newStreamState);
+    }
+    
+    return newState;
+};
+
+var _initQueueService = function() {
+    const storageParams = parse(process.env.AzureWebJobsStorage);
+    QueueService = azureStorage.createQueueService(
+        storageParams.AccountName, 
+        storageParams.AccountKey, 
+        storageParams.AccountName + '.queue.core.windows.net');
+    return QueueService;
+};
+
+var _getQueueService = function() {
+    return (QueueService) ? QueueService : _initQueueService();
+};
+
+var _fetch = function(callback) {
+    var queueService = _getQueueService();
+    var options = {};
+    const queueName = _getStateQueueName();
+    
+    options.visibilityTimeout = 180;
+    
+    queueService.getMessage(queueName, options,
+        function(error, message) {
+            if (error && error.code === 'QueueNotFound') {
+                return callback(null, _getNewListState());
+            } else if (error) {
+                return callback(`Get state $error`);
+            } else if (!message) {
+                // Another instance of a function is running.
+                return callback(`Singleton protection $error`);
```

Comment:
Do we have guarantee that we won't stuck here because Queue exist, but by some reason we lost all the messages there?

(notsureif)

Reply:
Yes we do. If a function which checks out the state from the queue crashes then the state message will be available in the queue when `visibilityTimeout` expires. So the following function invocations  will pick it up again.